### PR TITLE
Switch DisruptionBudget api from bool to int allowed disruptions

### PR DESCRIPTION
**Pull request by mwielgus**
Currently there is a single bool in disruption budget api that denotes whether 1 pod can be deleted or not. Every time a pod is deleted the apiserver filps the bool to false and the disruptionbudget controller sets it to true if more deletions are allowed. This works but it is far from optimal when the user wants to delete multiple pods (for example, by decreasing replicaset size from 10000 to 8000).
This PR changes bool to int which contains a number of pods that can be deleted at once.

cc: @davidopp @mml @wojtek-t @fgrzadkowski

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34546)

### CODE REVIEW

**Comment by kargakis** [comment]
You need to update validation to disallow negative values in the new field.

**Comment by mwielgus** [comment]
@kargakis  Validation added.

**Comment by mml** [comment]
The int is more flexible, but the bool has the property that should the disruption controller crash or simply be "behind", the largest number of evictions incorrectly approved is 1 per PDB.

I suggest that if we switch this to an int, we also have a flag that controls the highest value the controller is willing to set, and consider shipping with a default value of 1.  Users who want better performance should understand that they are taking on a certain amount of risk when they raise the flag.

This seems fairly important since the whole point of this system is safety.  Throughput or efficiency are probably properties admins would like to maximize, but not at the possible expense of the safety they sought in the first place.

**Comment by davidopp** [comment]
@mml I think PDB controller going down is OK, since the API server synchronously does
```
newPDB.Status.PodDisruptionsAllowed--
```
when it deletes the pod. Does that sound reasonable?

But this got me wondering about the whole mechanism by which PDB controller updates the status. It seems like there's maybe something broken here. Let's assume for discussion the current implementation. Say the API server deletes a pod in response to eviction subresource. Right before or after the deletion (which is it?), API server sets PodDisruptionsAllowed=false. The PDB controller sees (via watch) the update to the PDB, which triggers it to recompute the number of healthy pods and possibly set PodDisruptionsAllowed=true. Let me know if any of that is wrong.

Assuming that's right, it seems that we are relying on 
(1) PDB controller will have the current state of the pods when it processes the PDB, in particular it will have already seen the deletion of the pod
(2) the way PDB controller computes number of healthy pods will consider a pod that is in graceful termination (but not actually deleted) to not be healthy. (In fact everywhere we say "deleted" we really mean "DeletionTimestamp has been set or the pod is gone from etcd.)

I think (1) can't be guaranteed because PDB controller processes different objects in parallel, and (2) may not be guaranteed either because the code looks like 

```
		for _, c := range pod.Status.Conditions {
			if c.Type == api.PodReady && c.Status == api.ConditionTrue {
				currentHealthy++
			}
		}
```
and I'm worried that pod in graceful deletion will still count as "currentHealthy"?

If (1) is violated then it seems PDB controller could re-set PodDisruptionsAllowed back to "true" an unlimited number of times (say the watch on pods blocks, so PDB controller never sees the deletions, while the watch on PDB objects works fine). And if (2) is violated, then PDB controller could re-set PodDisruptionsAllowed back to "true" an unlimited number of times, though for a different reason (it's getting all the pod updates, but it doesn't think any pods have actually been evicted until 30 seconds later or whatever, when it discovers that it has allowed them all to be evicted).

@mwielgus had suggested that API server attach to the PDB status a list of pods whose deletion it has approved, and PDB controller would clear a pod from the list once it has incorporated that pod into its count of *un*healthy pods. I think this might address both of these concerns (and also yours?).

**Comment by mwielgus** [comment]
There will always be the rist that that disruption budget may be violated due to crashes. It might be a philosophical topic but i see being off by 1 problematic to the same degree as being off by, lets say, 5. The most vulnerable to disruptions are apps that have only 2-3 replicas (or 1 but they most likely will have disruption budget = 0 all the time). 

Anyway, @mml where would you like to add this flag? To the controller or disruption budget definition?

**Comment by mwielgus** [comment]
@davidopp Adding a list of  pods won't help in case of controller/master crash. Imagine a situation when the controller dies for 10 min. Before the failure all 10 pods controlled by the disruption budget were present. Disruption budget requires that 4 pods are always present. When the controller is down 6 of these pods are stopped. So the status of the disruption budget indicates that 6 pods can be deleted while none can actually be delete. If there was bool only one pod could be deleted.

**Comment by mml** [comment]
> I think PDB controller going down is OK, since the API server synchronously does
> 
> newPDB.Status.PodDisruptionsAllowed--
> when it deletes the pod. Does that sound reasonable?

@mwielgus is correct.  Here's a re-statement of the scenario.

1. Controller sees 10,000 ready of 8,000 required.  Sets `allowed` to 2,000.
2. Controller goes down.  No one notices.
3. Something happens: rack switch crash, power outage, cluster shrinks, whatever.  Now, fewer than 8,000 pods are healthy.
4. Eviction requests come in for 2,000 evictions and the API server dutifully decrements `allowed` one at a time.
5. Now, fewer than 6,000 pods are healthy.

Another thing that can make the number stale: if the replica set scale changes and the budget is percentage-based.

**Comment by mml** [comment]
@davidopp for your second question, I think I see what you mean with respect to (1) and (2).  Stated differently, we approve an eviction, decrement the counter, but the pod remains in a state where it appears healthy.  If the controller is running and the grace period is sufficiently long (a minute is probably enough), we could easily "safely" evict every single pod even if the budget allows for only 1 eviction, as long as the evicting agent was aggressive about retries.

The idea of tracking which pods have been approved for eviction so that they aren't counted as "healthy" is a good one.  We probably want those approvals to eventually expire.  We can calculate a safe expiration time when we approve: `current_time + grace_period + fudge_factor`.

**Comment by mml** [comment]
> Anyway, @mml where would you like to add this flag? To the controller or disruption budget definition?

I think an actual command-line flag on the controller-manager makes the most sense.  Any admin who wants to raise this value should start from a place where they have some kind of maintenance throughput problem to fix, and understand the risks.

**Comment by davidopp** [comment]
Ah, got it. Yeah, I think that regardless of whether we use a bool or an int, we need to make it clear that only on a best-effort basis do we take into account pods that are unavailable due to reasons other than eviction via eviction subresource. In the original design doc I left open the question of whether we should take into account actual state at all and I hadn't really thought through all of the implications of using actual state. (There's really no way to provide a perfectly strong guarantee when using actual state for this featuer anyway, because even if the PDB controller were consulted synchronously before the API responded "OK" to an eviction request, pods may die one nanosecond before the eviction request is received.)

Anyway, I think the outcome of that tangent is that we definitely need to add the evicted pods to PDBStatus and never consider such pods healthy. This can be done in a followup PR but is needed for 1.5.

As for this PR -- I think making the max allowed `PodDisruptionsAllowed` be a flag to the controller manager makes sense. (And I agree with defaulting it to 1.)

**Comment by davidopp** [comment]
And actually, since we also want to move this feature to beta in 1.5, I think you should change this PR to not make the change in the v1alpha1 version of PDB, but instead fork a new version of the API group (create `pkg/apis/policy/v1beta1`) and make the change to DisruptionBudgetStatus there. 

Whether we should keep support for the alpha version of the resource (`pkg/apis/policy/v1alpha1`) is a separate question. It would be nice to, since some people might have started using it in 1.4, but we make no guarantees about alpha stuff staying around so technically we can just remove it.

**Comment by mwielgus** [comment]
@davidopp @mml Does it mean this PR has your LGTM? (after fixing, probably minor, issues with build).

**Comment by davidopp** [comment]
The basic idea is fine, but it doesn't make sense to change the field in the alpha representation. You should fork a beta version and make the change there.

**Comment by mml** [comment]
@mwielgus I think you should add the flag in this PR as well, but the approach looks fine.

**Comment by mwielgus** [comment]
Let's limit the scope of the PR to just API. I will do the flag in the following one.

File: pkg/apis/policy/v1alpha1/conversion.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		return err
+	}
+	return nil
+}
+
+func Convert_v1alpha1_PodDisruptionBudgetStatus_To_policy_PodDisruptionBudgetStatus(in *PodDisruptionBudgetStatus, out *policy.PodDisruptionBudgetStatus, s conversion.Scope) error {
+	if in.PodDisruptionAllowed {
+		out.PodDisruptionsAllowed = 1
+	} else {
+		out.PodDisruptionsAllowed = 2
```
Comment (wojtek-t) [inline pkg/apis/policy/v1alpha1/conversion.go]: Should this be 2? Not 0?

Comment (mwielgus) [inline pkg/apis/policy/v1alpha1/conversion.go]: Fixed.

**Comment by wojtek-t** [comment]
```k8s.io/kubernetes/pkg/api TestRoundTripTypes 4.48s``` is failing - can you please test first?

Also - REALLY do NOT apply "retest-not-required". It is large non-trivial PR.

File: pkg/api/testing/fuzzer.go [hunk 2]
```diff
@@ -538,6 +539,10 @@ func FuzzerFor(t *testing.T, version unversioned.GroupVersion, src rand.Source)
 			// Set the bytes field on the RawExtension
 			r.Raw = bytes
 		},
+		func(s *policy.PodDisruptionBudgetStatus, c fuzz.Continue) {
+			c.FuzzNoCustom(s) // fuzz self without calling this function again
+			s.PodDisruptionsAllowed = c.Rand.Intn(2)
+		},
```
Comment (wojtek-t) [inline pkg/api/testing/fuzzer.go]: OK - this commit looks ok. Can you please squash it with the first one?

File: pkg/apis/policy/install/install.go [hunk 3]
```diff
@@ -22,18 +22,20 @@ import (
...
 	"k8s.io/kubernetes/pkg/apis/policy/v1alpha1"
+	"k8s.io/kubernetes/pkg/apis/policy/v1beta1"
 )
 
 func init() {
 	if err := announced.NewGroupMetaFactory(
 		&announced.GroupMetaFactoryArgs{
 			GroupName:                  policy.GroupName,
-			VersionPreferenceOrder:     []string{v1alpha1.SchemeGroupVersion.Version},
+			VersionPreferenceOrder:     []string{v1alpha1.SchemeGroupVersion.Version, v1beta1.SchemeGroupVersion.Version},
```
Comment (wojtek-t) [inline pkg/apis/policy/install/install.go]: What is the order within this slice? The last one is the most preferred one? If not, the order should probably be different.

File: pkg/controller/disruption/disruption_test.go [hunk 4]
```diff
@@ -53,23 +54,25 @@ func (ps *pdbStates) Get(key string) policy.PodDisruptionBudget {
...
 	}
 }
 
-func (ps *pdbStates) VerifyDisruptionAllowed(t *testing.T, key string, disruptionAllowed bool) {
+func (ps *pdbStates) VerifyDisruptionAllowed(t *testing.T, key string, disruptionsAllowed int32) {
 	pdb := ps.Get(key)
-	if pdb.Status.PodDisruptionAllowed != disruptionAllowed {
-		t.Fatalf("PodDisruptionAllowed mismatch for PDB %q.  Expected %v but got %v.", key, disruptionAllowed, pdb.Status.PodDisruptionAllowed)
+	if pdb.Status.PodDisruptionsAllowed != disruptionsAllowed {
+		debug.PrintStack()
```
Comment (wojtek-t) [inline pkg/controller/disruption/disruption_test.go]: Do we still need it? Or is it just for debugging?

**Review by wojtek-t — COMMENTED** [review]
Few minor comments. Other than that lgtm

File: pkg/api/testing/fuzzer.go (hunk 2 above)
Comment (mwielgus) [inline pkg/api/testing/fuzzer.go]: Ack

File: pkg/apis/policy/install/install.go (hunk 3 above)
Comment (mwielgus) [inline pkg/apis/policy/install/install.go]: Switched. Batch has v1 , v2alpha1 order here. The field is not documented, although switching it seems reasonable.

File: pkg/controller/disruption/disruption_test.go (hunk 4 above)
Comment (mml) [inline pkg/controller/disruption/disruption_test.go]: Yeah.  We should implement a wrapper around t.Fatalf() that uses runtime.Caller() to print the right thing.

**Comment by caesarxuchao** [comment]
The integration test failure message:
```
storage codec doesn't seem to match given type: Internal type not encodable: converting (v1beta1.PodDisruptionBudget).v1beta1.PodDisruptionBudgetStatus to (v1alpha1.PodDisruptionBudget).v1alpha1.PodDisruptionBudgetStatus: PodDisruptionAllowed not present in src
```
It looks like the codec thinks v1alpha1 is the internal version, but I ran the tests locally with `go test -tags=integration`, the tests passed. I'll continue debugging this later today.

**Comment by mwielgus** [comment]
@caesarxuchao Is the api enabled if you run this in that way? make test-integration passes some extra arguments.

**Comment by mwielgus** [comment]
@caesarxuchao  @lavalamp  It seems that the problem is with testapi. It stores api group definitions in a map which is keyed by a just a group, without a version. So if 2 versions are available the only the first one "wins" (groups are processed in reverse order). It creates confusion later when a different api version is to be used.

**Comment by caesarxuchao** [comment]
This line might be related: 
https://github.com/kubernetes/kubernetes/blob/master/test/integration/framework/master_utils.go#L332. I'll debug it later today.

**Comment by caesarxuchao** [comment]
In the integration test, we got a verioning codec wrapped around a versioning codec (https://github.com/kubernetes/kubernetes/blob/master/test/integration/framework/master_utils.go#L332.), which seems wrong. The outer layer codec encodes PDB to v1beta1, passed it to the inner layer codec, which tries to encode to v1alpha1, and thus the error. I need time to think about how to fix it. The problem is isolated to integration tests.

**Comment by mwielgus** [comment]
Closing in favor of #35567.

**Pull request closed without merging**